### ISSUE TITLE
Build/runtime error with infi.systray

### ISSUE THREAD

**Issue by Athanasius** [py2exe/py2exe#88]
We've been trying to get Windows system tray icon support into our project using `infi.systray`.  It works, but we're having trouble getting py2exe to package it properly.

Our issue is https://github.com/EDCD/EDMarketConnector/issues/1020

I've made an example code repo at https://github.com/Athanasius/infi-systray_py2exe

The `main` branch of that tries to do the bare minimum and gets the error as shown in https://github.com/Athanasius/infi-systray_py2exe/blob/main/sample_systray-bare-setup.log

The `setup-pkg_resources` branch then attempts to include `pkg_resources` but gets the error as shown in https://github.com/Athanasius/infi-systray_py2exe/blob/setup-pkg_resources/sample_systray-with-pkg_resources.log

Adding `'packages'` to py2exe `packages` option gets us a little further... but then can't import other things.  This is where I realised that `pkg_resources` includes a copy of `packages` in `pkg_resources._vendor`, but:

1. That isn't recognised by py2exe as a package (`__path__` is missing?)
2. Doesn't work if included as a module instead.

I realise this whole thing is because pkg_resources and packages are doing their own module importing directly, which makes it tricky for py2exe to find the correct files to include in the build.

**Comment by albertosottile** [comment]
Looks like `setuptools` is doing some shenanigans in the structure of `pkg_resources`. I will try to write a hook for it.

**Comment by albertosottile** [comment]
I tried to fix the bug in cb8282323d6743e2b34c385d354247b2e719f2a1. You can find wheels that include the tentative fix on [GitHub Actions](https://github.com/py2exe/py2exe/actions/runs/854725060).

@Athanasius if you could download the wheels you need and test if they fix the issue also for you, that would be great.

**Comment by Athanasius** [comment]
Thanks, I'll take a look at this tomorrow.

**Comment by Athanasius** [comment]
It appears to be working with that wheel.  I can't reproduce the same error, the sample_systray.exe runs without logging anything, and the system tray icon menu functions.

However I do still need to explicitly:

```python
            'includes': [
                'pkg_resources',
            ],
```
is this expected ?

**Comment by albertosottile** [comment]
This is expected with the current code, but perhaps an hook for `infi` should be added, since this dependency is not very clear... 

I added one in the latest build, could you download the updated wheels from [here](https://github.com/py2exe/py2exe/actions/runs/861686328) and try again? Thanks.

**Comment by Athanasius** [comment]
I'll try that out tomorrow.  Thanks for the work :) .

**Comment by Athanasius** [comment]
Yup, that latest .whl appears to work perfectly without manually specifying a `pkg_resources` include.

I look forward to the next stable release which includes this :) .

Thanks!

**Comment by albertosottile** [comment]
@Athanasius I apologize for the huge delay, but I just found the time to release https://github.com/py2exe/py2exe/releases/tag/v0.10.4.1, which as a matter of fact includes just this fix. I hope you will be able to integrate it in your projects, from now on. Thanks for your patience.

**Comment by Athanasius** [comment]
Yeah, sorry about going silent on this, we've been on a bit of a hiatus from most development.

I've just tested, and now I changed the setup.py we had in use to *not* exclude pkg_resources, this is indeed packaging correctly using released py2exe `0.10.4.1`.

**Comment by albertosottile** [comment]
I doubt that the fix for #126 somehow caused this, as it is only adding new files in the bundle. It is much more likely that this "feature" from `pkg_resources` broke during the 0.11 transition. Anyway, I will look at this as soon as I have some spare time.

**Comment by Athanasius** [comment]
> I doubt that the fix for #126 somehow caused this, as it is only adding new files in the bundle. It is much more likely that this "feature" from `pkg_resources` broke during the 0.11 transition. Anyway, I will look at this as soon as I have some spare time.

Our application works with:

```
0$ pip freeze | egrep '^(semantic-version|infi.systray|py2exe)'
infi.systray==0.1.12
py2exe==0.11.0.1
semantic-version==2.8.5 # It's 2.9.0 of this that changed its `__init__.py` and caused the issue in #126 
```
So it's some change after that py2exe version that is causing this.

**Comment by albertosottile** [comment]
What are your versions of Python and `setuptools`?

**Comment by albertosottile** [comment]
This does not add up, I am unable to reproduce. I took the sample you posted for this issue (https://github.com/Athanasius/infi-systray_py2exe/blob/main/sample_systray.py) and added `import semantic_version` in line 4.

Then, I packed it with the wheels from the `egginfo` branch (latest build: https://github.com/py2exe/py2exe/actions/runs/1834917831).  With this configuration, I am able to pack the bundle and also run it. 

Here is what I have:
```
C:\py2exe\tests\functional\infi>python --version
Python 3.10.2

C:\py2exe\tests\functional\infi>pip freeze
cachetools==5.0.0
future==0.18.2
infi.systray==0.1.12
pefile==2021.9.3
py2exe @ file:///C:/py2exe/py2exe-0.11.1.0-cp310-cp310-win_amd64.whl
semantic-version==2.9.0

pip-22.0.3 setuptools-60.8.2 wheel-0.37.1
```

Unfortunately, I cannot integrate this test in CI as the systray package will never work, and I do not have a local 32bit test environment.

My only suggestion at this point is to update `setuptools` and try again...

**Comment by Athanasius** [comment]
I'm also at a loss.  I just did, in a 'Git Bash' window unless otherwise specified:

1. Create a whole new directory
2. `git clone https://github.com/Athanasius/infi-systray_py2exe.git`
3. `cd infi-systray_py2exe/`
4. `deactivate` # to get back to base python 3.10.2 active, confirmed with `which python`.
5. `python -m venv venv`
6. `. venv/Scripts/activate`
7. Confirm `which python` points into the new venv.
8. `pip freeze` - it's empty
9. `pip install -r requirements.txt`
10. `pip freeze`:
    ```
    cachetools==5.0.0
    future==0.18.2
    infi.systray==0.1.12
    pefile==2021.9.3
    py2exe==0.11.0.1
    ```

11. Download the `py2exe-0.11.1.0-cp310-cp310-win32.whl.zip` from https://github.com/py2exe/py2exe/actions/runs/1834917831 - unzip the `.whl` from it.  This is:
    ```
    0$ sha256sum ~/Downloads/Devel/Python/py2exe/py2exe-0.11.1.0-cp310-cp310-win32.whl
    0705e3c5e511d6e83b2a8447b5b3e9d5d3fc91c8cecdbd95d79c7e9b9f0dc71d */c/Users/Athan/Downloads/Devel/Python/py2exe/py2exe-0.11.1.0-cp310-cp310-win32.whl
    ```

12. `pip install ~/Downloads/Devel/Python/py2exe/py2exe-0.11.1.0-cp310-cp310-win32.whl`
13. Now **in a Windows cmd.exe**:
    1. `cd` into the same path
    2. `venv\Scripts\deactivate.bat` # to ensure back to the base python
    3. `venv\Scripts\activate.bat` # to definitely use the local venv python
    4. `python -c "import sys ; print (sys.path)"` to confirm it's using this venv's python.
    5. `python setup.py py2exe`
    6. `cd dist`
    7. `sample_systray.exe` -> popup about output to the `sample_systray.log` file:
        ```
        Traceback (most recent call last):
          File "sample_systray.py", line 55, in <module>
          File "sample_systray.py", line 50, in main
          File "sample_systray.py", line 23, in __init__
        ModuleNotFoundError: No module named 'infi'
        ```
Output from the py2exe build was:

```
(venv) C:\Users\Athan\Documents\Devel\tmp\infi-systray_py2exe>python setup.py py2exe
C:\Users\Athan\Documents\Devel\tmp\infi-systray_py2exe\setup.py:1: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
  from distutils.core import setup
running py2exe

  25 missing Modules
  ------------------
? __builtin__                         imported from pkg_resources._vendor.pyparsing
? __main__                            imported from bdb, pdb, pkg_resources
? _frozen_importlib                   imported from importlib, importlib.abc, zipimport
? _frozen_importlib_external          imported from importlib, importlib._bootstrap, importlib.abc, zipimport
? _posixshmem                         imported from multiprocessing.resource_tracker, multiprocessing.shared_memory
? _winreg                             imported from pkg_resources._vendor.appdirs, platform
? asyncio.DefaultEventLoopPolicy      imported from -
? collections.Iterable                imported from pkg_resources._vendor.pyparsing
? collections.MutableMapping          imported from pkg_resources._vendor.pyparsing
? com.sun                             imported from pkg_resources._vendor.appdirs
? com.sun.jna                         imported from pkg_resources._vendor.appdirs
? com.sun.jna.platform                imported from pkg_resources._vendor.appdirs
? dummy.Process                       imported from multiprocessing.pool
? infi.systray                        imported from __main__
? java.lang                           imported from platform
? ordereddict                         imported from pkg_resources._vendor.pyparsing
? org.python.core                     imported from copy, pickle
? os.path                             imported from ctypes._aix, distutils.file_util, os, pkg_resources, pkgutil, py_compile, sysconfig, tracemalloc, unittest, unittest.util
? pep517                              imported from importlib.metadata
? pkg_resources.extern.appdirs        imported from pkg_resources
? pkg_resources.extern.packaging      imported from pkg_resources
? readline                            imported from cmd, code, pdb
? resource                            imported from test.support
? win32api                            imported from pkg_resources._vendor.appdirs
? win32com.shell                      imported from pkg_resources._vendor.appdirs
Building 'dist\sample_systray.exe'.
Add resource PYTHONSCRIPT/1(4337 bytes) to dist\sample_systray.exe
Building shared code archive 'dist\library.zip'.
Add Loader for _bz2.pyd to dist\library.zip
Add Loader for _tkinter.pyd to dist\library.zip
Add Loader for _overlapped.pyd to dist\library.zip
Add Loader for _uuid.pyd to dist\library.zip
Add Loader for _elementtree.pyd to dist\library.zip
Add Loader for _testcapi.pyd to dist\library.zip
Add Loader for _ctypes.pyd to dist\library.zip
Add Loader for _lzma.pyd to dist\library.zip
Add Loader for select.pyd to dist\library.zip
Add Loader for _asyncio.pyd to dist\library.zip
Add Loader for _decimal.pyd to dist\library.zip
Add Loader for unicodedata.pyd to dist\library.zip
Add Loader for pyexpat.pyd to dist\library.zip
Add Loader for _queue.pyd to dist\library.zip
Add Loader for _socket.pyd to dist\library.zip
Add Loader for _hashlib.pyd to dist\library.zip
Add Loader for _multiprocessing.pyd to dist\library.zip
Add Loader for _ssl.pyd to dist\library.zip
Add Loader for _testinternalcapi.pyd to dist\library.zip
Copy c:\program files (x86)\python310-32\python310.dll to dist
Copy C:\Program Files (x86)\Python310-32\DLLs\_bz2.pyd to dist\_bz2.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_lzma.pyd to dist\_lzma.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\unicodedata.pyd to dist\unicodedata.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_hashlib.pyd to dist\_hashlib.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_decimal.pyd to dist\_decimal.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_socket.pyd to dist\_socket.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\select.pyd to dist\select.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\pyexpat.pyd to dist\pyexpat.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_ssl.pyd to dist\_ssl.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_queue.pyd to dist\_queue.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_elementtree.pyd to dist\_elementtree.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_testcapi.pyd to dist\_testcapi.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_ctypes.pyd to dist\_ctypes.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_tkinter.pyd to dist\_tkinter.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_testinternalcapi.pyd to dist\_testinternalcapi.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_asyncio.pyd to dist\_asyncio.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_multiprocessing.pyd to dist\_multiprocessing.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_overlapped.pyd to dist\_overlapped.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_uuid.pyd to dist\_uuid.pyd
Copy DLL C:\Program Files (x86)\Python310-32\DLLs\tk86t.dll to dist\
Copy DLL C:\Program Files (x86)\Python310-32\DLLs\tcl86t.dll to dist\
Copy DLL C:\Program Files (x86)\Python310-32\DLLs\libssl-1_1.dll to dist\
Copy DLL C:\Program Files (x86)\Python310-32\DLLs\libcrypto-1_1.dll to dist\
Copy DLL C:\Program Files (x86)\Python310-32\DLLs\libffi-7.dll to dist\

(venv) C:\Users\Athan\Documents\Devel\tmp\infi-systray_py2exe>cd dist

(venv) C:\Users\Athan\Documents\Devel\tmp\infi-systray_py2exe\dist>sample_systray.exe

(venv) C:\Users\Athan\Documents\Devel\tmp\infi-systray_py2exe\dist>type sample_systray.log
Traceback (most recent call last):
  File "sample_systray.py", line 55, in <module>
  File "sample_systray.py", line 50, in main
  File "sample_systray.py", line 23, in __init__
ModuleNotFoundError: No module named 'infi'

(venv) C:\Users\Athan\Documents\Devel\tmp\infi-systray_py2exe\dist>
```

**Comment by Athanasius** [comment]
If this is an x86 (32-bit) versus amd64 (64-bit) thing then that's an issue.  Our application still has some ctypes code in it that will need updating to run as 64-bit, which is a dark art when no-one still involved has the knowledge of Windows 'C' APIs to be sure what the types of things need to be.

**Comment by Athanasius** [comment]
And after a `pip install --upgrade setuptools` in the venv for the test above it works (well, then there's some actual infi.systray issues, but that's nothing to do with py2exe).  Let me double-check all of that with our actual app now.

**Comment by Athanasius** [comment]
No, I was blind, you said:

> pip-22.0.3 setuptools-60.**8.2** wheel-0.37.1

when I'd last reported:

> venv/Lib/site-packages/setuptools-60.**6.0**.dist-info/

I think it's time we start listing setuptools in our requirements-dev.txt then dependabot will prod us to update.

**Comment by Athanasius** [comment]
If it wasn't clear, now I made sure the venv has `setuptools==60.8.2` the build *of our app* actually works, as did the example systray code under those circumstances.  I'll take a look at your verbose log comment in a moment....

**Comment by Athanasius** [comment]
I think the py2exe output from infi.systray might have only been missing due to the wrong setuptools version.  That last run with 60.8.2:

```
(venv) C:\Users\Athan\Documents\Devel\tmp\infi-systray_py2exe>python setup.py py2exe
running py2exe

  31 missing Modules
  ------------------
? __builtin__                         imported from pkg_resources._vendor.pyparsing
? __main__                            imported from bdb, pdb, pkg_resources
? _frozen_importlib                   imported from importlib, importlib.abc, zipimport
? _frozen_importlib_external          imported from importlib, importlib._bootstrap, importlib.abc, zipimport
? _manylinux                          imported from pkg_resources._vendor.packaging._manylinux
? _posixshmem                         imported from multiprocessing.resource_tracker, multiprocessing.shared_memory
? _winreg                             imported from pkg_resources._vendor.appdirs, platform
? asyncio.DefaultEventLoopPolicy      imported from -
? collections.Iterable                imported from pkg_resources._vendor.pyparsing
? collections.MutableMapping          imported from pkg_resources._vendor.pyparsing
? com.sun                             imported from pkg_resources._vendor.appdirs
? com.sun.jna                         imported from pkg_resources._vendor.appdirs
? com.sun.jna.platform                imported from pkg_resources._vendor.appdirs
? dummy.Process                       imported from multiprocessing.pool
? java.lang                           imported from platform
? ordereddict                         imported from pkg_resources._vendor.pyparsing
? org.python.core                     imported from copy, pickle
? os.path                             imported from ctypes._aix, distutils.file_util, os, pkg_resources, pkgutil, py_compile, sysconfig, tracemalloc, unittest, unittest.util
? pep517                              imported from importlib.metadata
? pkg_resources.extern.appdirs        imported from pkg_resources
? pkg_resources.extern.importlib_resources imported from pkg_resources._vendor.jaraco.text
? pkg_resources.extern.jaraco.context imported from pkg_resources._vendor.jaraco.text
? pkg_resources.extern.jaraco.functools imported from pkg_resources._vendor.jaraco.text
? pkg_resources.extern.jaraco.text    imported from pkg_resources
? pkg_resources.extern.more_itertools imported from pkg_resources._vendor.jaraco.functools
? pkg_resources.extern.packaging      imported from pkg_resources
? pkg_resources.extern.pyparsing      imported from pkg_resources._vendor.packaging.markers, pkg_resources._vendor.packaging.requirements
? readline                            imported from cmd, code, pdb
? resource                            imported from test.support
? win32api                            imported from pkg_resources._vendor.appdirs
? win32com.shell                      imported from pkg_resources._vendor.appdirs
Building 'dist\sample_systray.exe'.
Add resource PYTHONSCRIPT/1(4337 bytes) to dist\sample_systray.exe
Building shared code archive 'dist\library.zip'.
Add Loader for _overlapped.pyd to dist\library.zip
Add Loader for unicodedata.pyd to dist\library.zip
Add Loader for _asyncio.pyd to dist\library.zip
Add Loader for _ssl.pyd to dist\library.zip
Add Loader for _multiprocessing.pyd to dist\library.zip
Add Loader for _testinternalcapi.pyd to dist\library.zip
Add Loader for _queue.pyd to dist\library.zip
Add distribution metadata for package infi.systray from c:\users\athan\documents\devel\tmp\infi-systray_py2exe\venv\lib\site-packages\infi.systray-0.1.12.dist-info
Add Loader for _decimal.pyd to dist\library.zip
Add Loader for _tkinter.pyd to dist\library.zip
Add Loader for _bz2.pyd to dist\library.zip
Add Loader for _socket.pyd to dist\library.zip
Add Loader for _uuid.pyd to dist\library.zip
Add Loader for _testcapi.pyd to dist\library.zip
Add Loader for select.pyd to dist\library.zip
Add Loader for _hashlib.pyd to dist\library.zip
Add Loader for _ctypes.pyd to dist\library.zip
Add Loader for _lzma.pyd to dist\library.zip
Add empty __init__ for implicit namespace package infi to dist\library.zip
Add Loader for _elementtree.pyd to dist\library.zip
Add Loader for pyexpat.pyd to dist\library.zip
Copy c:\program files (x86)\python310-32\python310.dll to dist
Copy C:\Program Files (x86)\Python310-32\DLLs\_bz2.pyd to dist\_bz2.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_lzma.pyd to dist\_lzma.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\unicodedata.pyd to dist\unicodedata.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_hashlib.pyd to dist\_hashlib.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_decimal.pyd to dist\_decimal.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_socket.pyd to dist\_socket.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\select.pyd to dist\select.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\pyexpat.pyd to dist\pyexpat.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_ssl.pyd to dist\_ssl.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_queue.pyd to dist\_queue.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_elementtree.pyd to dist\_elementtree.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_testcapi.pyd to dist\_testcapi.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_ctypes.pyd to dist\_ctypes.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_tkinter.pyd to dist\_tkinter.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_testinternalcapi.pyd to dist\_testinternalcapi.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_asyncio.pyd to dist\_asyncio.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_multiprocessing.pyd to dist\_multiprocessing.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_overlapped.pyd to dist\_overlapped.pyd
Copy C:\Program Files (x86)\Python310-32\DLLs\_uuid.pyd to dist\_uuid.pyd
Copy DLL C:\Program Files (x86)\Python310-32\DLLs\tcl86t.dll to dist\
Copy DLL C:\Program Files (x86)\Python310-32\DLLs\tk86t.dll to dist\
Copy DLL C:\Program Files (x86)\Python310-32\DLLs\libffi-7.dll to dist\
Copy DLL C:\Program Files (x86)\Python310-32\DLLs\libssl-1_1.dll to dist\
Copy DLL C:\Program Files (x86)\Python310-32\DLLs\libcrypto-1_1.dll to dist\
```

**Comment by albertosottile** [comment]
That would make sense, given that there are a bunch of try/except around calls to `pkg_resources` in the code that handle those additions. However, I tried to downgrade my `setuptools` to find out the minimum supported version (to make those log lines disappear) and I was not able to do so (in my x86_64 test setup). I went down to 58.1.0 and I still have a functional bundle. 

I find unlikely the architecture has an influence here, as `setuptools` is a pure Python package, but I still do not understand completely what happened in your system (a note though: in general, building from a virtual environment is not supported, as `distutils` does not like that. I am quite surprised you managed to got this far in this condition).

Anyway, the idea was to add a minimum version of `setuptools` in the requirements of `py2exe` but, because I cannot find one, I have abandoned it and I will still just recommend to use this software with "the latest version of `pip`/`setuptools`/`wheel` available".

Because you were able to bundle your app, I will now close this issue and focus on releasing the fix for #126. Thanks again for your support!

**Comment by Athanasius** [comment]
Well, I use a virtual environment for actual development and 'in-person' testing because otherwise I'd end up with a mess of modules needed for other things affecting this, and modules only needed for this affecting other things.

For the actual GitHub builds it just uses their `windows-2019` and `actions/setup-python`.

There's a separate thought about `distutils` which I'll go check for existing issues on, and if none exist I'll open one.
